cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,213 @@
 
 #pragma once
 
+#include <cstdint>
 #include <cudf/strings/string_view.cuh>
+#include <hash/hash_constants.hpp>
+
+#include "cuda_runtime_api.h"
+#include "cudf/types.hpp"
+#include "driver_types.h"
+#include "vector_types.h"
 
 using hash_value_type = uint32_t;
 
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Core MD5 algorith implementation. Processes a single 512-bit chunk,
+ * updating the hash value so far. Does not zero out the buffer contents.
+ */
+void CUDA_DEVICE_CALLABLE md5_hash_step(md5_intermediate_data* hash_state,
+                                        md5_hash_constants_type const* hash_constants,
+                                        md5_shift_constants_type const* shift_constants)
+{
+  uint32_t A = hash_state->hash_value[0];
+  uint32_t B = hash_state->hash_value[1];
+  uint32_t C = hash_state->hash_value[2];
+  uint32_t D = hash_state->hash_value[3];
+
+  for (unsigned int j = 0; j < 64; j++) {
+    uint32_t F;
+    uint32_t g;
+    switch (j / 16) {
+      case 0:
+        F = (B & C) | ((~B) & D);
+        g = j;
+        break;
+      case 1:
+        F = (D & B) | ((~D) & C);
+        g = (5 * j + 1) % 16;
+        break;
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    uint32_t buffer_element_as_int;
+    thrust::copy_n(thrust::seq,
+                   hash_state->buffer + g * 4,
+                   4,
+                   reinterpret_cast<uint8_t*>(&buffer_element_as_int));
+    F = F + A + hash_constants[j] + buffer_element_as_int;
+    A = D;
+    D = C;
+    C = B;
+
+    uint32_t shift = shift_constants[((j / 16) * 4) + (j % 4)];
+    B              = B + ((F << shift) | (F >> (32 - shift)));
```

Comment:
use funnel_shift here. 
Check other bit intrinsic functions too.  
https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html

Reply:
I substituted `__funnelshift_l` in -- replacing lines 78-79 with `B = B + __funnelshift_l(F, F, shift_constants[((j / 16) * 4) + (j % 4)]);`  but I'm a little hesitant to commit and push the changes because it more than tripled my build time. Any idea why this would cause such a massive jump?

Reply:
Besides, build time. Does it improve performance?
Is it built only for your GPU architecture or all GPU architectures?

Reply:
I was building for multiple architectures, but I don't think I was comparing apples to apples because I was using ninja to build cudf. In a few tests adding and reverting changes, the compile times seem nearly identical.